pkg/cloudprovider/openstack.go

```diff
@@ -237,45 +219,126 @@ func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
-				// Try this 10 times, but if this operation fails more than that, then user intervention is needed or
-				// the upper layer must call ReleasePrivateIP (because if the neutron port exists and holds
-				// a reservation, then the assign step will not continue after step 2).
-				var errRelease error
-				var releaseStatus string
-				for i := 0; i < 10; i++ {
-					errRelease = o.releaseNeutronIPAddress(*unboundPort, serverID)
-					// If the release operation was successful, then we are done.
-					if errRelease == nil {
-						releaseStatus = "Released neutron port reservation."
-						break
-					}
-					// Otherwise store the error message and retry.
-					releaseStatus = fmt.Sprintf("Could not release neutron port reservation after %d tries, err: %q", i+1, errRelease)
-				}
-				return fmt.Errorf("could not allow IP address %s on port %s, err: %q. %s", ip.String(), serverPort.ID, err, releaseStatus)
+			// Otherwise store the error message and retry.
+			releaseStatus = fmt.Sprintf("Could not release neutron port reservation after %d tries, err: %q", i+1, errRelease)
+		}
+		return fmt.Errorf("could not allow IP address %s on port %s, err: %q. %s", ip.String(), matchingPort.ID, err, releaseStatus)
+	}
+	// 4) Return nil to indicate success if steps 2 and 3 passed.
+	return nil
+}
+
+// MovePrivateIP implements moving the IP from one node to another to serve cases like a failover.
+// What's different from calling ReleasePrivateIP followed by AssignPrivateIP is that the reservation
+// Neutron port will not get deleted - MovePrivateIP will only change the allowed_address_pairs on the node's
+// ports to remove IP from nodeToDel and add it to nodeToAdd.
+func (o *OpenStack) MovePrivateIP(ip net.IP, nodeToAdd, nodeToDel *corev1.Node) error {
+	if nodeToAdd == nil || nodeToDel == nil {
+		return fmt.Errorf("invalid nil pointer provided for node when trying to move IP %s", ip.String())
+	}
+
+	// List all ports that are attached to this server.
+	serverID, err := getNovaServerIDFromProviderID(nodeToDel.Spec.ProviderID)
+	if err != nil {
+		return err
+	}
+	serverPorts, err := o.listNovaServerPorts(serverID)
+	if err != nil {
+		return err
+	}
+
+	// Loop over all ports that are attached to this nova instance.
+	for _, serverPort := range serverPorts {
+		if isIPAddressAllowedOnNeutronPort(serverPort, ip) {
+			if err = o.unallowIPAddressOnNeutronPort(serverPort.ID, ip); err != nil {
+				return err
 			}
-			// 4) Return nil to indicate success if steps 2 and 3 passed.
-			return nil
 		}
 	}
 
-	// 5) The IP address does not fit in any of the attached networks' subnets.
-	return fmt.Errorf("could not assign IP address %s to node %s", ip, node.Name)
+	subnet, port, err := o.findAssignSubnetAndPort(ip, nodeToAdd)
+	if err != nil {
+		return err
+	}
+
+	// This call is to double-check if the reservation port exists, function
+	// will return okay if the port with that IP already exists in the subnet.
```

Comment:
... will return okay if the port with that IP already exists in the subnet, otherwise it will create the missing reservation port.